src/object/cli_obj.c

```diff
@@ -4535,14 +4535,14 @@ obj_size_fetch_cb(const struct dc_object *obj, struct obj_auxi_args *obj_auxi)
  * For that case, we duplicate the sgls and make its iov_buf_len = iov_len.
  */
 static int
-obj_update_sgls_dup(struct obj_auxi_args *obj_auxi, daos_obj_update_t *args)
+obj_sgls_dup(struct obj_auxi_args *obj_auxi, daos_obj_update_t *args, bool update)
```

Comment:
Just curious, why do we need this duplication.  daos client stack can not handle iov_buf == NULL/iov_lent = 0?

Reply:
will hit some failures in mercury if with that kind of iov

Reply:
Just curious can we fix this to allow some iov_buf in sgls to be iov_buf == NULL/iov_len = 0  in mercury? or I miss sth. @soumagne

Reply:
> Just curious can we fix this to allow some iov_buf in sgls to be iov_buf == NULL/iov_len = 0 in mercury? or I miss sth. @soumagne

To be clear, I am not very sure if the failure is caused by mercury.
The failure reported in daos log is -
ERROR: daos_engine:1 02/05-01:57:03.55 boro-41 DAOS[3615093/3/112] object ERR  src/object/srv_obj.c:449 bulk_transfer_sgl() Remote bulk isn't large enough. local_sz:43692, remote_sz:98304, remote_off:87380, DER_OVERFLOW(-1017): 'Data too long for defined data type or buffer size'
ERROR: daos_engine:1 02/05-01:57:03.55 boro-41 DAOS[3615093/3/112] object ERR  src/object/srv_obj.c:578 obj_bulk_transfer() bulk_transfer_sgl i 0, skip_nr 0 failed, DER_OVERFLOW(-1017): 'Data too long for defined data type or buffer size'
...
and 
ERROR: daos_engine:1 02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] bio  ERR  src/bio/bio_buffer.c:475 copy_one() Invalid iov[0] 0/0 1
ERROR: daos_engine:1 02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] object ERR  src/object/srv_obj.c:1653 obj_local_rw_internal() 2251804108652548.1.1.1 data transfer failed, dma 0: DER_INVAL(-1003): 'Invalid parameters'
02/05-01:48:02.12 boro-35 DAOS[4138276/3/16267] object ERR  src/object/srv_obj.c:152 obj_rw_complete() 2251804108652548.1.1.1 Fetch end failed: DER_INVAL(-1003): 'Invalid parameters'
.

This patch just dup+removes the (NULL, 0) iov that is the simplest fix, the dup is not needed for common case in critical path.

Reply:
what test caused such errors? I don't quite see how the errors are related to this PR.

Reply:
> what test caused such errors? I don't quite see how the errors are related to this PR.

the changed test case in this PR ("daos_test -i -u 12") can reproduce the failure.

Reply:
@wangdi1 sorry I completely missed your question, yes it is allowed to pass a bulk handle that has NULL/0 lengths, this was requested long time ago.

Reply:
Thanks @soumagne , so @liuxuezhao  this failure might cause by some offset mis-calculation inside daos stack? or I miss sth?